apps/indexer/src/api/services/historical-balances/historical-balances.service.ts

```diff
@@ -0,0 +1,163 @@
+import {
+  Address,
+  createPublicClient,
+  http,
+  PublicClient,
+  isAddress,
+  parseAbi,
+} from "viem";
+import { readContract, multicall } from "viem/actions";
+
+import { DaoIdEnum } from "@/lib/enums";
+import { CONTRACT_ADDRESSES } from "@/lib/constants";
+import { getChain } from "@/lib/utils";
+import { env } from "@/env";
+
+export interface HistoricalBalance {
+  address: Address;
+  balance: bigint;
+  blockNumber: number;
+  tokenAddress: Address;
+}
+
+export interface HistoricalBalancesRequest {
+  addresses: Address[];
+  blockNumber: number;
+  daoId: DaoIdEnum;
+}
+
+export class HistoricalBalancesService {
+  private client: PublicClient;
+
+  constructor() {
+    const { CHAIN_ID: chainId, RPC_URL: rpcUrl } = env;
+    const chain = getChain(chainId);
+
+    if (!chain) {
+      throw new Error(`Chain with ID ${chainId} not found`);
+    }
+
+    this.client = createPublicClient({
+      chain,
+      transport: http(rpcUrl),
+    });
+  }
+
+  /**
+   * Fetches historical balances for multiple addresses at a specific block
+   * Uses Multicall3 for efficient batch queries when available (block 19+)
+   * Falls back to individual calls for earlier blocks
+   */
+  async getHistoricalBalances({
+    addresses,
+    blockNumber,
+    daoId,
+  }: HistoricalBalancesRequest): Promise<HistoricalBalance[]> {
+    const tokenAddress = this.getTokenAddress(daoId);
+
+    if (!tokenAddress) {
+      throw new Error(`Token address not found for DAO: ${daoId}`);
+    }
+
```

Comment:
how is this being handled by the API?

Reply:
The input of the API gets only the daoId(UNI, ENS, OP), and then get the token address from the CONTRACT_ADDRESSES, so this line will never be accessed, since the wrong daoId triggers a validation error:
`{"error":"Validation Error","message":"Validation error: Expected 'UNI' | 'ENS' | 'ARB', received null at \"daoId\"","details":[{"expected":"'UNI' | 'ENS' | 'ARB'","received":"null","code":"invalid_type","path":["daoId"],"message":"Expected 'UNI' | 'ENS' | 'ARB', received null"}]}`